backends/graphics/openglsdl/openglsdl-graphics.cpp

```diff
@@ -352,37 +348,54 @@ bool OpenGLSdlGraphicsManager::loadVideoMode(uint requestedWidth, uint requested
+	float ratio = (float)desktopRes.width() / (float)desktopRes.height();
+
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+	bool _isMaximized = (SDL_GetWindowFlags(_window->getSDLWindow()) & SDL_WINDOW_MAXIMIZED);
+	if (_isMaximized && ConfMan.hasKey("window_maximized_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("window_maximized_height", Common::ConfigManager::kApplicationDomain)) {
 		// Set the window size to the values stored when the window was maximized
-		// for the last time. We also need to reset any scaling here.
+		// for the last time.
 		requestedWidth  = ConfMan.getInt("window_maximized_width", Common::ConfigManager::kApplicationDomain);
 		requestedHeight = ConfMan.getInt("window_maximized_height", Common::ConfigManager::kApplicationDomain);
 
-	} else if (ConfMan.hasKey("last_window_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("last_window_height", Common::ConfigManager::kApplicationDomain)) {
-		// Restore previously stored window dimensions.
+	} else if (!_isMaximized && ConfMan.hasKey("last_window_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("last_window_height", Common::ConfigManager::kApplicationDomain)) {
+		// Load previously stored window dimensions.
 		requestedWidth  = ConfMan.getInt("last_window_width", Common::ConfigManager::kApplicationDomain);
 		requestedHeight = ConfMan.getInt("last_window_height", Common::ConfigManager::kApplicationDomain);
 
 	} else {
 		// Set the basic window size based on the desktop resolution
 		// since we have no values stored, e.g. on first launch.
-		requestedWidth  = desktopRes.width()  * 0.3f;
-		requestedHeight = desktopRes.height() * 0.4f;
-
-		// Apply scaler
-		requestedWidth  *= _graphicsScale;
-		requestedHeight *= _graphicsScale;
+		requestedWidth  = MAX<uint>(desktopRes.width() / 2, MIN<uint>(640, desktopRes.width()));
+		requestedHeight = MIN<uint>(requestedWidth * 3 / 4, desktopRes.height());
 
 		// Save current window dimensions
 		ConfMan.setInt("last_window_width", requestedWidth, Common::ConfigManager::kApplicationDomain);
 		ConfMan.setInt("last_window_height", requestedHeight, Common::ConfigManager::kApplicationDomain);
 		ConfMan.flushToDisk();
 	}
 
-	// Determine current aspect ratio
+#else
+		// Set the basic window size based on the desktop resolution
+		// since we cannot reliably determine the current window state
+		// on SDL1.
+		requestedWidth  = MAX<uint>(desktopRes.width() / 2, MIN<uint>(640, desktopRes.width()));
+		requestedHeight = MIN<uint>(requestedWidth * 3 / 4, desktopRes.height());
+
+#endif
+
+	// In case we detected an increase in the requested window dimensions (e.g. when launching
+	// a game in 800x600 while having a smaller screen size stored in the configuration file),
+	// we override the window dimensions with the "real" request.
+	if ((requestedWidth < _lastRequestedWidth || requestedHeight < _lastRequestedHeight) && ConfMan.getActiveDomain()) {
```

Comment:
```suggestion
	if ((requestedWidth < _lastRequestedWidth * _graphicsScale || requestedHeight < _lastRequestedHeight * _graphicsScale) && ConfMan.getActiveDomain()) {
```

This only matters for game with a width <= 320 (as then `_graphicsScale` is 2) and when the window is bigger than the requested size but smaller than double the requested size. In such a case we could say that the current size is fine (and not double the size), which is what your code is doing. But then we probably want to also change the two lines below to not use the `_graphicsScale`.

The current inconsistency using `_graphicsScale` in the two lines below, but not in the `if` condition, mean a strange behaviour where a smaller current window can result in a bigger window.
| current size | requested size | final size |
|-------------|----------------|----------|
|   310x200    |      320x200      | 640x400|
|   340x220    |      320x200     | 340x220|

So either we use the `_graphicsScale` in both places, or in neither.

Reply:
Very nice, I was confused already why the scaling randomly didn't work properly for the 320x200 games. Turns out that the missing `_graphicsScale` in the comparison was indeed the culprit here.

Fixed by using `_graphicsScale` at both instances now.